### PR TITLE
[WIP] Clone transaction before committing to wallet

### PR DESCRIPTION
**Pull request by sqrrm**
Created a new PR after the discussions in https://github.com/bisq-network/bitcoinj/pull/18

### CODE REVIEW

**Comment by sqrrm** [comment]
I think we should get this in for the release after next, so no rush but good to run in testnet for a bit.

File: core/src/main/java/org/bitcoinj/wallet/Wallet.java [hunk 1]
```diff
@@ -1698,6 +1698,11 @@ public void receivePending(Transaction tx, @Nullable List<Transaction> dependenc
                 return;
             if (isTransactionRisky(tx, dependencies) && !acceptRiskyTransactions) {
                 // isTransactionRisky already logged the reason.
+
+                // Clone transaction to avoid multiple wallets pointing to the same transaction. This can happen when
+                // two wallets depend on the same transaction.
+                tx = tx.getParams().getDefaultSerializer().makeTransaction(tx.bitcoinSerialize());
+
```
[the remark this answers is not in the document]
Comment (sqrrm) [reply core/src/main/java/org/bitcoinj/wallet/Wallet.java]: It stores it in riskDropped and returns, that's why it needs to be cloned here.

[the remark this answers is not in the document]
Comment (oscarguindzberg) [reply core/src/main/java/org/bitcoinj/wallet/Wallet.java]: ah, you are right.

**Comment by oscarguindzberg** [comment]
utACK

**Comment by ManfredKarrer** [comment]
@sqrrm 
I discussed the issue with @oscarguindzberg 
I think it is too risky to clone every tx at a low level. We are not clear what can be the consequences as the pointers to the connected outputs (from the inputs) are not not cloned and might cause issues at some use cases.
Another point is that conceptually a single tx shared between wallets is probably the correct design and the follow up problems should rather be fixed instead of copying txs. This all requires quite a bit of work and analysis and as long we don't have a strong use case where we run into problems I prefer to postpone that to a later point of time.
So I would suggest to close the PR.

**Comment by sqrrm** [comment]
@ManfredKarrer I think it's quite clear that having the same tx instance in multiple wallets in bitcoinj is not how it's intended. When transactions are received from blocks they're cloned, see AbstractBlockChain:882. I think the reason they're not cloned when received from the broadcast callback is an oversight from the wallet developer.

The connected outputs are only set after a tx enters the wallet from what I understand. This also fits with how the external txs are cloned when added to multiple wallets.

I can still understand your hesitation to merge this too soon, but I think the conclusions are wrong regarding the risks and how the bitcoinj wallet works. Since we're already doing the cloning when publishing from the BSQ wallet it's not extremely urgent to fix this, but as mentioned before, I think it's worth fixing before going live on mainnet.

**Comment by oscarguindzberg** [comment]
Related https://github.com/bitcoinj/bitcoinj/pull/1553

**Pull request closed without merging**